internal/safari_webdriver.py

```diff
@@ -0,0 +1,1144 @@
+                # Figure out the native viewport size
+                size = self.execute_js('return [window.innerWidth, window.innerHeight]')
+                logging.debug(size)
+                if size and len(size) == 2:
+                    task['actual_viewport'] = {'width':size[0], 'height':size[1]}
+                    if 'adjust_viewport' in job and job['adjust_viewport']:
+                        delta_x = max(task['width'] - size[0], 0)
+                        delta_y = max(task['height'] - size[1], 0)
+                        if delta_x or delta_y:
+                            width = task['width'] + delta_x
+                            height = task['height'] + delta_y
+                            logging.debug('Resizing browser to %dx%d', width, height)
+                            self.set_window_size(width, height)
+                # Wait for the browser startup to finish
+                DesktopBrowser.wait_for_idle(self)
+        except Exception as err:
+            logging.error('Error starting Safari')
+            task['error'] = 'Error starting Safari: {0}'.format(err.__str__())
+
+    def get_pref_value(self, value):
+        """Convert a JSON pref value to Python"""
+        str_match = re.match(r'^"(.*)"$', value)
+        if value == 'true':
+            value = True
+        elif value == 'false':
+            value = False
+        elif re.match(r'^[\d]+$', value):
+            value = int(value)
+        elif str_match:
+            value = str_match.group(1)
+        else:
+            value = None
+        return value
+
+    def configure_prefs(self):
+        """Load the prefs file and configure them through Webdriver"""
+        prefs = {}
+        prefs_file = os.path.join(os.path.abspath(os.path.dirname(__file__)),
+                                  'support', 'Safari', 'profile', 'prefs.js')
+        with open(prefs_file) as f_in:
+            for line in f_in:
+                matches = re.search(r'user_pref\("([^"]+)",[\s]*([^\)]*)[\s]*\);', line)
+                if matches:
+                    key = matches.group(1).strip()
+                    value = self.get_pref_value(matches.group(2).strip())
+                    if value is not None:
+                        prefs[key] = value
+
+    def driver_set_pref(self, key, value):
+        """Set a Safari pref at runtime using Webdriver"""
+        # TODO(wuwang): add set_pref for Safari
+        pass
+
+    def set_window_size(self, width, height):
+        """Position the window"""
+        self.webdriver.set_window_rect(x=0, y=0, height=height, width=width)
+
+    def disconnect_driver(self):
+        """Disconnect Webdriver"""
+        if self.extension_id is not None and self.addons is not None:
+            try:
+                self.addons.uninstall(self.extension_id)
+            except Exception:
+                logging.error('Error removing addons')
+            self.extension_id = None
```

Comment:
It doesn't look like extension_id is used anywhere in the code.  Is this left over?

Reply:
Same goes for self.addons

Reply:
done